### PR TITLE
port: Several Adaptive parity fixes and additions

### LINKED ISSUE #3040: [PORT] Add support for choosing assignment or operations
https://github.com/microsoft/botbuilder-dotnet/pull/4855

This adds support for choosing an assignment or operation. For example, "help" for your name will ask if you want to assign help to your name or get help.

Improved robustness for handling op + property when values are required.

Fixed some bugs around global operations and choosing between properties.

### LINKED ISSUE #3297: port: Clean-up OnChooseEntity and OnChooseProperty schemas and classes (#5180)
The changes in [Clean-up OnChooseEntity and OnChooseProperty schemas and classes (#5180)](https://github.com/microsoft/botbuilder-dotnet/pull/5180) may need to be ported to maintain parity with `microsoft/botbuilder-dotnet`.

<blockquote>
Fixes #5177 

## Description
In writing the documentation for OnChooseEntity and OnChooseProperty I realized the operation was missing from OnChooseEntity and none of the properties for OnChooseProperty were used or even made sense.  This updated both the class and the schema.  These events are only fired if there is a schema and the only current source of schema is form generation which has not been released or documented yet.

## Testing
Improved the OnChooseEntity test to make sure filter properties work.
</blockquote>

Please review and, if necessary, port the changes.

### LINKED ISSUE #3320: port: Rename entity to value in form related triggers (#5228)
The changes in [Rename entity to value in form related triggers (#5228)](https://github.com/microsoft/botbuilder-dotnet/pull/5228) may need to be ported to maintain parity with `microsoft/botbuilder-dotnet`.

<blockquote>
Both operation and property are also entities so it was strange to have an entity property--renaming it to value makes the intent clearer.
</blockquote>

Please review and, if necessary, port the changes.

### PR DESCRIPTION
**Pull request by mdrichardson**
Fixes #3040
Fixes #3320
Fixes #3297

Note: Some of the above issue's parents touch some test files not changed in this PR because they don't exist. I'll address this in https://github.com/microsoft/botbuilder-js/issues/3611

## Description

This PR started as just addressing #3297, but since there were no tests related to Entities in JS yet, a lot of the methods in AdaptiveDialog and related test methods were broken and this PR quickly expanded to incorporating the other two along with many fixes just to make the OnChooseEntity tests pass. I re-wrote a lot of them so they're now more of a 1:1 match with .NET; it might be easier to [compare against that](https://github.com/microsoft/botbuilder-dotnet/blob/main/libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/AdaptiveDialog.cs) than to read the diff when reviewing.

### CODE REVIEW

File: libraries/botbuilder-dialogs-adaptive-testing/src/mocks/mockLuisRecognizer.ts [hunk 1]
```diff
@@ -69,7 +69,7 @@ export class MockLuisRecognizer extends Recognizer {
         this._recognizer = recognizer;
         this._responseDir = path.join(resourceDir, 'cachedResponses', name);
         if (!fs.existsSync(this._responseDir)) {
-            fs.mkdirSync(this._responseDir);
+            fs.mkdirSync(this._responseDir, { recursive: true });
```
Comment (mdrichardson) [inline libraries/botbuilder-dialogs-adaptive-testing/src/mocks/mockLuisRecognizer.ts:72]: Note: This was preventing the `cachedResponses` child folders from getting created if `cachedResponses\` didn't already exist.

File: libraries/botbuilder-dialogs-adaptive/src/actions/setProperty.ts [hunk 2]
```diff
@@ -127,6 +127,6 @@ export class SetProperty<O extends object = {}> extends Dialog<O> implements Set
      * @returns A `string` representing the compute Id.
      */
     protected onComputeId(): string {
-        return `SetProperty[${this.value.toString()}]`;
+        return `SetProperty[${this.property.toString()}]`;
```
Comment (mdrichardson) [inline libraries/botbuilder-dialogs-adaptive/src/actions/setProperty.ts:130]: Note: for [parity w/ .NET](https://github.com/microsoft/botbuilder-dotnet/blob/main/libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Actions/SetProperty.cs#L112)

File: libraries/botbuilder-dialogs-adaptive/src/input/ask.ts [hunk 3]
```diff
@@ -113,4 +117,11 @@ export class Ask extends SendActivity implements AskConfiguration {
         result.status = DialogTurnStatus.completeAndWait;
         return result;
     }
+
+    protected onComputeId(): string {
+        if (this.activity instanceof ActivityTemplate) {
+            return `Ask[${StringUtils.ellipsis(this.activity.template.trim(), 30)}]`;
+        }
+        return `Ask[${StringUtils.ellipsis(this.activity && this.activity.toString().trim(), 30)}]`;
```
Comment (mdrichardson) [inline libraries/botbuilder-dialogs-adaptive/src/input/ask.ts:125]: Note: Parity w/ .NET. This would previously show as `SendActivity[...` instead of `Ask[...` which confused me greatly during debugging.

File: libraries/botbuilder-dialogs-declarative/src/defaultLoader.ts [hunk 4]
```diff
@@ -39,6 +39,9 @@ export class DefaultLoader implements CustomDeserializer<Configurable, Record<st
                 }
                 instance[`${key}`] = converter.convert(value);
             } else {
+                if (key === 'schema' && typeof value === 'string') {
+                    value = JSON.parse(this._resourceExplorer.getResource(value).readText());
+                }
                 instance[`${key}`] = value;
             }
```
Comment (mdrichardson) [inline libraries/botbuilder-dialogs-declarative/src/defaultLoader.ts:47]: Note: Some of the `.dialog` test files were written like:

```json
{
   <normal .dialog JSON>,
   "schema": "someSchemaFile.json"
}
```

In .NET, it's smart enough to see that `someSchemaFile.json` and appropriately reads the file and deserializes it into an object:

```json
{
   <normal .dialog JSON>,
   "schema": {
		<someSchemaFiles's JSON>
	}
}
```

However, in JS this stays as the string, `someSchemaFile.json` and the file would never get read, so you couldn't have this nested `"schema"` property.

**TL;DR** I don't know if this is the best place to do this, but it made sense to me. In .NET, [`JObjectConverter`](https://github.com/microsoft/botbuilder-dotnet/blob/main/libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Converters/JObjectConverter.cs) eventually calls [`GetRefTarget()`](https://github.com/microsoft/botbuilder-dotnet/blob/main/libraries/Microsoft.Bot.Builder.Dialogs.Declarative/Resources/ResourceExplorer.cs#L576) and knows to read the file and convert it.

File: libraries/botbuilder-dialogs/src/dialogManager.ts [hunk 5]
```diff
@@ -78,7 +78,7 @@ export class DialogManager extends Configurable {
         if (rootDialog) {
             this.rootDialog = rootDialog;
         }
-        this._dialogStateProperty = dialogStateProperty || 'DialogStateProperty';
+        this._dialogStateProperty = dialogStateProperty || 'DialogState';
```
Comment (mdrichardson) [inline libraries/botbuilder-dialogs/src/dialogManager.ts:81]: Big thanks to @joshgummersall for debugging this one, that prevented `MemoryAssertions` from properly inspecting the dialog state. [This matches .NET.](https://github.com/microsoft/botbuilder-dotnet/blob/main/libraries/Microsoft.Bot.Builder.Dialogs/DialogManager.cs#L34)

File: libraries/botbuilder-dialogs-adaptive/etc/botbuilder-dialogs-adaptive.api.md [hunk 6]
```diff
@@ -1071,10 +1071,11 @@ export class OnCancelDialog extends OnDialogEvent {
 
 // @public
 export class OnChooseEntity extends OnDialogEvent {
-    constructor(property?: string, entity?: string, actions?: Dialog[], condition?: string);
+    constructor(property?: string, value?: string, operation? string, actions?: Dialog[], condition?: string);
```
Comment (joshgummersall) [inline libraries/botbuilder-dialogs-adaptive/etc/botbuilder-dialogs-adaptive.api.md:1074]: It would be potentially useful to add something to CI that regenerates these files and makes sure there are no diffs (to make sure we keep them up to date).

File: libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts [hunk 7]
```diff
@@ -1038,165 +1028,313 @@ export class AdaptiveDialog<O extends object = {}> extends DialogContainer<O> im
...
+                            // Handle operator with no children.
+                            this.expandEntity(op, null, instance, root, op, property, turn, text, entityToInfo);
+                        } else if (isProperty) {
+                            // Handle property with no children.
+                            this.expandEntity(property, null, instance, root, op, property, turn, text, entityToInfo);
                         }
                     } else {
-                        infos.push(info);
+                        this.expandEntityObject(
+                            entity as Record<string, unknown[]>,
```
Comment (joshgummersall) [inline libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts:1121]: Whenever I'm forced to use `as T` style casting, I like to leave a comment explaining why.

File: libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts [hunk 8]
```diff
@@ -1038,165 +1028,313 @@ export class AdaptiveDialog<O extends object = {}> extends DialogContainer<O> im
...
                     }
+                } else if (isOp) {
+                    // Handle global operator with no children in model.
+                    this.expandEntity(op, null, instance, root, op, property, turn, text, entityToInfo);
                 }
+            });
+        }
+    }
+
+    private stripProperty(name: string) {
```
Comment (joshgummersall) [inline libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts:1140]: Return type

File: libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts [hunk 9]
```diff
@@ -1038,165 +1028,313 @@ export class AdaptiveDialog<O extends object = {}> extends DialogContainer<O> im
...
+        return name.endsWith(this.propertyEnding) ? name.substring(0, name.length - this.propertyEnding.length) : name;
+    }
+
+    private expandEntity(
```
Comment (joshgummersall) [inline libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts:1144]: Return type

File: libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts [hunk 10]
```diff
@@ -1038,165 +1028,313 @@ export class AdaptiveDialog<O extends object = {}> extends DialogContainer<O> im
...
 
-                // Eventually this could be passed in
-                info.priority = info.role ? 0 : 1;
-                info.coverage = (info.end - info.start) / text.length;
-                if (propertyName) {
-                    // Add property information to entities
-                    if (propertyName.start < info.start) {
-                        info.start = propertyName.start;
-                    }
+    private expandEntities(
```
Comment (joshgummersall) [inline libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts:1060]: Return type

File: libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts [hunk 11]
```diff
@@ -1038,165 +1028,313 @@ export class AdaptiveDialog<O extends object = {}> extends DialogContainer<O> im
         return entityToInfo;
     }
 
-    /**
-     * @private
-     */
-    private expandEntity(
-        entry: any,
-        metaData: any,
+    private expandEntityObject(
```
Comment (joshgummersall) [inline libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts:1031]: Return type

File: libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts [hunk 12]
```diff
@@ -1038,165 +1028,313 @@ export class AdaptiveDialog<O extends object = {}> extends DialogContainer<O> im
...
+        instance: Record<string, unknown>,
+        rootInstance: Record<string, unknown>,
+        op: string,
+        property: string,
+        turn: number,
+        text: string,
+        entityToInfo: Record<string, EntityInfo[]>
+    ) {
+        if (instance && rootInstance) {
+            if (!entityToInfo[name]) {
```
Comment (joshgummersall) [inline libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts:1156]: You can write these as `object[key] ??= defaultValue` in a single line.

File: libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts [hunk 13]
```diff
@@ -1038,165 +1028,313 @@ export class AdaptiveDialog<O extends object = {}> extends DialogContainer<O> im
...
-                    });
+    private candidates(
+        entities: NormalizedEntityInfos,
+        expected: string[],
+        lastEvent: string,
+        nextAssignment: EntityAssignment,
+        askDefault: Record<string, unknown>,
+        dialogDefault: Record<string, unknown>
+    ): EntityAssignment[] {
+        const globalExpectedOnly: string[] = (this.dialogSchema.schema[this.expectedOnlyKey] as string[]) || [];
```
Comment (joshgummersall) [inline libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts:1199]: `??` instead of `||`

File: libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts [hunk 14]
```diff
@@ -1238,47 +1383,53 @@ export class AdaptiveDialog<O extends object = {}> extends DialogContainer<O> im
...
+            if (propChoices != null) {
+                const entityPreferences = this.entityPreferences(propChoices);
+                let choices = perProperty[propChoices];
+
+                // Assume preference by order listed in mappings
+                // Alternatives would be to look at coverage or other metrics
+                for (const entity of entityPreferences) {
+                    let candidate: EntityAssignment;
+                    do {
+                        candidate = undefined;
```
Comment (joshgummersall) [inline libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts:1412]:
```typescript
candidate = choices.find(mapping => mapping.value.name === entity);
```

File: libraries/botbuilder-dialogs-adaptive/src/entityAssignmentComparer.ts [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+ */
+/**
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import { AdaptiveEvents } from './adaptiveEvents';
+import { EntityAssignment } from './entityAssignment';
+
+export class EntityAssignmentComparer {
+    private static eventPreference: string[] = [
```
Comment (joshgummersall) [inline libraries/botbuilder-dialogs-adaptive/src/entityAssignmentComparer.ts:12]: Generally, `tsc` can infer the type based on the type of the initial value. In those cases, prefer to omit the explicit type.

File: libraries/botbuilder-dialogs-adaptive/src/entityAssignmentComparer.ts [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+        AdaptiveEvents.assignEntity,
+        AdaptiveEvents.chooseProperty,
+        AdaptiveEvents.chooseEntity,
+    ];
+
+    constructor(private operationPreference: string[]) {}
```
Comment (joshgummersall) [inline libraries/botbuilder-dialogs-adaptive/src/entityAssignmentComparer.ts:18]: This value seems unused?

File: libraries/botbuilder-dialogs-adaptive/src/entityAssignments.ts [hunk 17]
```diff
@@ -11,28 +11,31 @@ import { EntityAssignment } from './entityAssignment';
...
  * @remarks When processing entities possible ambiguities are identified and when resolved they turn into assign events.
  * This tracking persists across multiple input utterances.
  */
-export class EntityAssignments {
-    /**
-     * Queue of entity assignments.
-     */
-    public assignments: Partial<EntityAssignment>[] = [];
+export class EntityAssignments implements EntityAssignmentsConfiguration {
+    constructor(public assignments: EntityAssignment[] = []) {}
```
Comment (joshgummersall) [inline libraries/botbuilder-dialogs-adaptive/src/entityAssignments.ts:24]: Can this be made `readonly`?

File: libraries/botbuilder-dialogs-adaptive/src/entityAssignments.ts [hunk 18]
```diff
@@ -11,28 +11,31 @@ import { EntityAssignment } from './entityAssignment';
...
 
     /**
      * Read entity event queue from memory.
      * @param actionContext Memory context.
      */
     public static read(actionContext: ActionContext): EntityAssignments {
-        let queues: EntityAssignments = actionContext.state.getValue(events);
-        if (!queues) {
-            queues = new EntityAssignments();
+        let queuesObject: EntityAssignments = actionContext.state.getValue(events);
```
Comment (joshgummersall) [inline libraries/botbuilder-dialogs-adaptive/src/entityAssignments.ts:31]: I think `getValue` supports a second `defaultValue` parameter which would make the `if` check unnecessary.

File: libraries/botbuilder-dialogs-adaptive/src/entityInfo.ts [hunk 19]
```diff
@@ -126,6 +126,24 @@ export class EntityInfo {
         );
     }
 
+    /**
+     * Returns true if entities share the same root.
+     * @param _this Source entity.
+     * @param entity Entity to compare.
+     */
+     public static sharesRoot(_this: Partial<EntityInfo>, entity: Partial<EntityInfo>): boolean {
```
Comment (joshgummersall) [inline libraries/botbuilder-dialogs-adaptive/src/entityInfo.ts:134]: Maybe call it `source` instead of `_this`

File: libraries/botbuilder-dialogs/src/dialogManager.ts (hunk 5 above)
Comment (joshgummersall) [reply libraries/botbuilder-dialogs/src/dialogManager.ts:81]: Should be `??` - glad to help as well!

File: libraries/botbuilder-dialogs-declarative/src/defaultLoader.ts [hunk 20]
```diff
@@ -39,6 +39,9 @@ export class DefaultLoader implements CustomDeserializer<Configurable, Record<st
                 }
                 instance[`${key}`] = converter.convert(value);
             } else {
+                if (key === 'schema' && typeof value === 'string') {
+                    value = JSON.parse(this._resourceExplorer.getResource(value).readText());
```
Comment (joshgummersall) [inline libraries/botbuilder-dialogs-declarative/src/defaultLoader.ts:43]: In order to best preserve existing behavior we should do this: 
- If `key` is `schema` and `value` is string (and possibly if `value.endsWith(.json)`
- Try to load file `value`, try to parse, if success then assign
- Otherwise, assign `value` as before

File: libraries/botbuilder-dialogs-adaptive/src/input/ask.ts [hunk 21]
```diff
@@ -49,6 +51,7 @@ export class Ask extends SendActivity implements AskConfiguration {
     public constructor(text?: string, expectedProperties?: ArrayExpression<string>) {
         super(text);
         this.expectedProperties = expectedProperties;
+        this.activity = new ActivityTemplate(text);
```
Comment (joshgummersall) [inline libraries/botbuilder-dialogs-adaptive/src/input/ask.ts:54]: Is `this.activity` defined on `SendActivity`? It feels like this ought to be logic that exists in the type that defines `this.activity`.

File: libraries/botbuilder-dialogs-adaptive/src/entityAssignment.ts [hunk 22]
```diff
@@ -5,9 +5,10 @@
  * Copyright (c) Microsoft Corporation. All rights reserved.
  * Licensed under the MIT License.
  */
+import { isEqual } from 'lodash';
 import { EntityInfo } from './entityInfo';
 
-export interface EntityAssignment {
+export interface EntityAssignmentConfiguration {
```
Comment (joshgummersall) [inline libraries/botbuilder-dialogs-adaptive/src/entityAssignment.ts:11]: Do we need an interface that's separate from the class?

File: libraries/botbuilder-dialogs-adaptive/src/entityAssignment.ts [hunk 23]
```diff
@@ -42,18 +43,78 @@ export interface EntityAssignment {
...
+    public raisedCount: number;
+
+    /**
+     * The expected properties when assignment was made.
+     */
+    public expectedProperties: string[];
+
+    constructor(assignment: Partial<EntityAssignmentConfiguration>) {
+        const newEntity: Partial<EntityAssignmentConfiguration> = {
+            alternative: assignment.alternative ?? undefined,
```
[the remark this answers is not in the document]
Comment (mdrichardson) [reply libraries/botbuilder-dialogs-adaptive/src/entityAssignment.ts:99]: Forgot to leave that as a comment. `undefined` is necessary because there are some areas where an equality check is done on two objects and `_.isEqual()` doesn't understand that this:

```js
const obj = {
	a: 'a'
}
```

should be considered equal to:

```js
```js
const obj = {
	a: 'a',
	b: undefined
}
```

Since the constructor accepts `Partial`s, I figured the best thing to do was to just ensure all properties are added in some way or another. Very open to other ideas.

File: libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts [hunk 24]
```diff
@@ -427,6 +442,7 @@ export class AdaptiveDialog<O extends object = {}> extends DialogContainer<O> im
         let state: AdaptiveDialogState = activeDialogState[this.adaptiveKey];
         if (!state) {
             state = { actions: [] };
+            activeDialogState[this.adaptiveKey] = state;
```
[the remark this answers is not in the document]
Comment (mdrichardson) [reply libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts:445]: [That's how it's done in .NET](https://github.com/microsoft/botbuilder-dotnet/blob/main/libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/AdaptiveDialog.cs#L341). I thought that was a little odd when I went over it.

File: libraries/botbuilder-dialogs-adaptive/src/entityAssignmentComparer.ts (hunk 16 above)
Comment (mdrichardson) [reply libraries/botbuilder-dialogs-adaptive/src/entityAssignmentComparer.ts:18]: That's only because I ported it over poorly :(

File: libraries/botbuilder-dialogs-adaptive/src/entityAssignments.ts (hunk 17 above)
Comment (mdrichardson) [reply libraries/botbuilder-dialogs-adaptive/src/entityAssignments.ts:24]: Negative. It gets modified outside of the class. Does so in .NET, too.

File: libraries/botbuilder-dialogs-adaptive/src/input/ask.ts (hunk 21 above)
Comment (mdrichardson) [reply libraries/botbuilder-dialogs-adaptive/src/input/ask.ts:54]: It is, so I'll remove. [.NET does the same thing](https://github.com/microsoft/botbuilder-dotnet/blob/main/libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Input/Ask.cs#L49) and I forgot to remove the change after debugging.

File: libraries/botbuilder-dialogs-adaptive/src/entityAssignment.ts (hunk 22 above)
Comment (mdrichardson) [reply libraries/botbuilder-dialogs-adaptive/src/entityAssignment.ts:11]: I did this so that we could deserialize `Partial` JSON objects into the class. I considered a constructor that took all arguments as optional, but that would've been 8 arguments and it seemed like a typed object might be better? I also needed to ensure that any properties not defined were explicitly set `undefined` as explained [here](https://github.com/microsoft/botbuilder-js/pull/3628#discussion_r621496959).

I went back and forth a bit about how best to do this and can't say I figured out the "right" solution. Happy to adjust if you think I should.

File: libraries/botbuilder-dialogs-declarative/src/defaultLoader.ts (hunk 20 above)
Comment (mdrichardson) [reply libraries/botbuilder-dialogs-declarative/src/defaultLoader.ts]: [This work?](https://github.com/microsoft/botbuilder-js/pull/3628/files#diff-7a0e9072626dfb54b4cacb48459f1ce6b0648e86e4894b87b2d5b5b1b6c960e8R42)

File: libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts [hunk 25]
```diff
@@ -53,6 +36,31 @@ import {
...
+import { BoolProperty } from './properties';
+import { RecognizerSet } from './recognizers';
+import { ValueRecognizer } from './recognizers/valueRecognizer';
+import { SchemaHelper } from './schemaHelper';
+import { FirstSelector, MostSpecificSelector } from './selectors';
+import { TriggerSelector } from './triggerSelector';
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+function isDialogDependencies(val: any): val is DialogDependencies {
+    return typeof ((val as unknown) as DialogDependencies).getDependencies == 'function';
```
Comment (joshgummersall) [inline libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts:62]: `===`

File: libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts [hunk 26]
```diff
@@ -419,17 +433,18 @@ export class AdaptiveDialog<O extends object = {}> extends DialogContainer<O> im
...
      * @param dc The [DialogContext](xref:botbuilder-dialogs.DialogContext) for the current turn of conversation.
      * @returns The child [DialogContext](xref:botbuilder-dialogs.DialogContext) or null if no [AdaptiveDialogState.actions](xref:botbuilder-dialogs-adaptive.AdaptiveDialogState.actions) are found for the given context.
      */
     public createChildContext(dc: DialogContext): DialogContext {
         const activeDialogState = dc.activeDialog.state;
-        let state: AdaptiveDialogState = activeDialogState[this.adaptiveKey];
+        const state: AdaptiveDialogState = activeDialogState[this.adaptiveKey];
         if (!state) {
-            state = { actions: [] };
+            activeDialogState[this.adaptiveKey] = { actions: [] };
```
Comment (joshgummersall) [inline libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts:444]: I don't think this actually propagates back to `state` (see [Typescript playground sample](https://www.typescriptlang.org/play?#code/MYewdgzgLgBA5gJxAVwA4C4YCUCmoEAmAPNAgJZhwA0MyYA1mCAO5gB8MAvDAN4C+AbgBQQ0JFj0cATy4wA5JKlzho8NBjQAhlByzEKVAG1FAXWFkAZjAAUAQi06AlLyEw38JGmPSTsnosw5ADdNABtkHDlBIT4RMQgQUJwAOlCQOGsHHEcBIA)).

```typescript
activeDialogStatel[this.adaptiveKey] ??= {actions: []};
const state = activeDialogState[this.adaptiveKey];
```

File: libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts [hunk 27]
```diff
@@ -1038,165 +1027,314 @@ export class AdaptiveDialog<O extends object = {}> extends DialogContainer<O> im
...
+        }
+    }
+
+    private stripProperty(name: string): string {
+        return name.endsWith(this.propertyEnding) ? name.substring(0, name.length - this.propertyEnding.length) : name;
+    }
+
+    private expandEntity(
+        name: string,
+        value: Record<string, unknown>,
```
Comment (joshgummersall) [inline libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts:1148]: [Here](https://github.com/microsoft/botbuilder-js/pull/3628/files#diff-a116dc3a22cc3932c498b73b60c065e4a035945b35eff59dc6130a086739015eR1116) this method is invoked with `value` set to `null`, `value: Record<string, unknown>` disallows `null` (`strict` mode would catch this, but we are in limbo at the moment). Add `| null` to this type just to be explicit.

File: libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts [hunk 28]
```diff
@@ -1038,165 +1027,314 @@ export class AdaptiveDialog<O extends object = {}> extends DialogContainer<O> im
...
+        if (instance && rootInstance) {
+            entityToInfo[name] ??= [];
+
+            const info: EntityInfo = {
+                whenRecognized: turn,
+                name,
+                value,
+                operation: op,
+                property,
+                start: <number>rootInstance.startIndex,
```
Comment (joshgummersall) [inline libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts:1166]: Are these `<T>` necessary?

File: libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts [hunk 29]
```diff
@@ -1038,165 +1027,314 @@ export class AdaptiveDialog<O extends object = {}> extends DialogContainer<O> im
...
+            if (!assignment.property) {
+                assignment.expectedProperties = expected;
             }
         }
 
-        return candidates;
+        return assignments;
     }
 
     /**
```
Comment (joshgummersall) [inline libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts:1317]: You could strip these `@private` JSDoc comments, they're not required anymore.

File: libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts [hunk 30]
```diff
@@ -1238,47 +1383,46 @@ export class AdaptiveDialog<O extends object = {}> extends DialogContainer<O> im
...
-            for (const entity of entityPreferences) {
-                let candidate: Partial<EntityAssignment>;
-                do {
-                    candidate = undefined;
-                    for (let i = 0; i < choices.length; i++) {
-                        const mapping = choices[i];
-                        if (mapping.entity.name == entity) {
-                            candidate = mapping;
-                            break;
+        const perProperty = candidates.reduce<{ [path: string]: EntityAssignment[] }>((accumulator, assignment): Record<
```
Comment (joshgummersall) [inline libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts:1388]: `reduce` is generally good about inferring types, are these explicit type params necessary? If anything, you should only have to provide `reduce<T>((acc, curr) => ..., defaultValue);` and the rest is inferred.

File: libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts [hunk 31]
```diff
@@ -1238,47 +1383,46 @@ export class AdaptiveDialog<O extends object = {}> extends DialogContainer<O> im
...
+            string,
+            EntityAssignment[]
+        > => {
+            const groupBy = assignment.property;
+            if (accumulator[groupBy]) {
+                accumulator[groupBy].push(assignment);
+            } else {
+                accumulator[groupBy] = [assignment];
+            }
+            return accumulator;
```
Comment (joshgummersall) [inline libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts:1398]: If you want to get fancy (I'm code golfing a bit here, so feel free to ignore this)

```typescript
const perProperty = candidates.reduce((acc, assignment) => ({
  ...acc,
  [assignment.property]: [...acc[assignment.property], assignment],
}), {}); 
```

File: libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts [hunk 32]
```diff
@@ -1298,136 +1442,144 @@ export class AdaptiveDialog<O extends object = {}> extends DialogContainer<O> im
...
+                this.addAssignment(choice, assignments);
+                choices = choices.filter((c): boolean => !EntityInfo.overlaps(c, choice.value));
             }
 
             existing.dequeue(actionContext);
         }
 
-        this.mergeAssignments(assignments, existing);
-        return Object.values(usedEntities);
+        // eslint-disable-next-line prettier/prettier
```
Comment (joshgummersall) [inline libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts:1554]: Why is this disabled?

File: libraries/botbuilder-dialogs-adaptive/src/entityAssignment.ts (hunk 23 above)
[the remark this answers is not in the document]
Comment (joshgummersall) [reply libraries/botbuilder-dialogs-adaptive/src/entityAssignment.ts:100]: I'm not in love with this code either, but your explanation is fair and makes sense. I would specifically cite `_.isEqual` behavior in the comment above as `undefined` and a key that doesn't exist should be semantically equivalent ([javascript.info is a good source of wisdom here](https://javascript.info/object#property-existence-test-in-operator)).

File: libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts (hunk 26 above)
Comment (mdrichardson) [reply libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts:444]: Correct. In this method, it doesn't need to since `state` isn't used in other blocks. I'll change the next `if` to an `else if` to make that more clear.

File: libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts (hunk 28 above)
Comment (mdrichardson) [reply libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts:1166]: Since `instance` and `rootInstance` are one of those weird nested objects, all of their properties are `unknown`. The linter complains if I remove the casting. Granted, at this point, we kind of know the shapes of `instance` and `rootInstance`; I just couldn't find a pre-defined type. This is currently [what .NET does](https://github.com/microsoft/botbuilder-dotnet/blob/main/libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/AdaptiveDialog.cs#L1184), too, not that that means we can't make it better. Let me know how you'd like me to proceed.

File: libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts (hunk 29 above)
Comment (mdrichardson) [reply libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts:1317]: Hallelujah!

File: libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts (hunk 30 above)
Comment (mdrichardson) [reply libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts:1388]: Not necessary. Removed.

File: libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts (hunk 32 above)
Comment (mdrichardson) [reply libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts:1554]: Not sure. Must've been a bad `ctrl+.` that I missed.

File: libraries/botbuilder-dialogs-adaptive/src/entityAssignment.ts (hunk 23 above)
[the remark this answers is not in the document]
Comment (mdrichardson) [reply libraries/botbuilder-dialogs-adaptive/src/entityAssignment.ts:100]: And just to hammer it home, it's not just `lodash`. Node's `util.isDeepStrictEqual()` and `assert.deepStrictEqual()` operate the same way. I considered writing a custom equality checker, but that seemed like although equally beneficial, it was more complex. Could go either way.

File: libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts [hunk 33]
```diff
@@ -1038,174 +1014,314 @@ export class AdaptiveDialog<O extends object = {}> extends DialogContainer<O> im
...
-                        entity: alternative,
-                        property: alternative.property,
-                        operation: alternative.operation,
-                        isExpected: expected.indexOf(alternative.property) >= 0,
-                    });
+        // Add default operations.
+        for (const assignment of assignments) {
+            if (!assignment.operation) {
+                // Assign missing operation.
+                if (lastEvent == AdaptiveEvents.chooseEntity && assignment.value.property === nextAssignment.property) {
```
Comment (joshgummersall) [inline libraries/botbuilder-dialogs-adaptive/src/adaptiveDialog.ts:1241]: `===` (you could search + replace in this file if there are any others that don't involve `null` or `undefined`)

**Review by joshgummersall — APPROVED** [review]
One final tiny suggestion. Nice work here!

File: libraries/botbuilder-dialogs-adaptive/src/entityAssignment.ts (hunk 23 above)
[the remark this answers is not in the document]
Comment (joshgummersall) [reply libraries/botbuilder-dialogs-adaptive/src/entityAssignment.ts:102]: I think it's fine, the comment clearly explains what's going on. All good!

File: libraries/botbuilder-dialogs-adaptive/src/entityAssignment.ts (hunk 22 above)
Comment (joshgummersall) [reply libraries/botbuilder-dialogs-adaptive/src/entityAssignment.ts:11]: Makes sense - let's keep brainstorming ways to express "an instance JSON parsed from a string that has extra computed properties" as it seems like a common pattern, and is something that isn't provided out of the box like it is in .NET. For now, this works.
